cmd/kubeadm/app/cmd/upgrade/common.go

```diff
@@ -95,12 +95,10 @@ func loadConfig(cfgPath string, kubeConfigPath string, skipComponentConfigs bool
 		return nil, false, err
 	}
 
-	// If there are kubeadm types (API group kubeadm.kubernetes.io) present, we need to keep the existing behavior
-	// here. Basically, we have to load all of the configs from the file and none from the cluster. Configs that are
-	// missing from the file will be automatically regenerated by kubeadm even if they are present in the cluster.
-	// The resulting configs overwrite the existing cluster ones at the end of a successful upgrade apply operation.
+	// If there are kubeadm types (API group kubeadm.kubernetes.io) present, we need to keep the existing behavior here.
+	// Component configs that are missing from the file will be automatically regenerated by kubeadm if they are not present in the cluster.
```

Comment:
oh, i forgot about this part.
currently, users can pass a --config to kubeadm upgrade that has the componentconfigs for kubelet and kube-proxy present, separated by ---.

this was done to allow component config upgrades, as there is no other UX for the user.
i was not a big fan of this when it was done, as it steps over the mutation on upgrade boundary (not desired).

IMO, upgrade's --config should not allow mixing UpgradeConfiguration with the component config types.
WDYT?

Reply:
more thoughts, by not allowing InitConfigiration on upgrade --config any more, we are effectively breaking users that pass the component config on upgrade.

perhaps we should not allow the component config on upgrade and that's all.
this might need some more discussion.

Reply:
> --config should not allow mixing UpgradeConfiguration with the component config types.

I don't have enough context on this, I am not sure is there any user are using the mixing configuraition. @SataQiu @pacoxu  thoughts?

> by not allowing InitConfigiration on upgrade --config any more, we are effectively breaking users that pass the component config +init configuration on upgrade.

- we should deprecate the use of `InitConfigiration` on upgrade, as there are some upgrade specific configuration which `InitConfigiration` cannot provide.
- Have a `InitConfigiration` in the upgrade data, built it from the current cluster and is immutable, while user's input of a Init config file means the upgrade might mutable via the new configuration.

Reply:
the whole logic is refactored.